eth/downloader/statesync.go

```diff
@@ -0,0 +1,331 @@
+		var deliverReqCh chan *stateReq
+		if len(finishedReqs) > 0 {
+			deliverReq = finishedReqs[0]
+			deliverReqCh = s.deliver
+		}
+
+		select {
+		// The stateSync lifecycle:
+		case next := <-d.stateSyncStart:
+			return next
+		case <-s.done:
+			return nil
+
+		// Send the next finished request to the current sync:
+		case deliverReqCh <- deliverReq:
+			finishedReqs = finishedReqs[1:]
+
+		// Handle incoming state packs:
+		case pack := <-d.stateCh:
+			req := activeReqs[pack.PeerId()]
+			if req == nil {
+				log.Debug("Unrequested node data", "peer", pack.PeerId(), "len", pack.Items())
+				continue
+			}
+			delete(activeReqs, pack.PeerId())
+			req.timer.Stop()
+			req.response = pack.(*statePack).states
+			finishedReqs = append(finishedReqs, req)
+
+		// Handle timed-out requests:
+		case req := <-timeout:
+			if activeReqs[req.peer.id] != req {
+				continue // ignore old timeout
+			}
+			finishedReqs = append(finishedReqs, req)
+			delete(activeReqs, req.peer.id)
+
+		// Track outgoing state requests:
+		case req := <-d.trackStateReq:
+			activeReqs[req.peer.id] = req
+			req.timer = time.AfterFunc(req.timeout, func() {
+				select {
+				case timeout <- req:
+				case <-cancelTimeout:
+				}
+			})
+		}
+	}
+}
+
+// stateSync schedules requests for downloading a particular state root.
+type stateSync struct {
+	d              *Downloader
+	sched          *state.StateSync
+	keccak         hash.Hash
+	tasksAvailable map[common.Hash]*stateTask
+
+	deliver    chan *stateReq
+	cancel     chan struct{}
+	cancelOnce sync.Once
+	done       chan struct{}
+	err        error // set after done is closed
+}
+
+type stateTask struct {
+	triedPeers map[string]struct{}
+}
+
+func newStateSync(d *Downloader, root common.Hash) *stateSync {
+	return &stateSync{
+		d:              d,
+		sched:          state.NewStateSync(root, d.stateDB),
+		keccak:         sha3.NewKeccak256(),
+		tasksAvailable: make(map[common.Hash]*stateTask),
+		deliver:        make(chan *stateReq),
+		cancel:         make(chan struct{}),
+		done:           make(chan struct{}, 1),
+	}
+}
+
+// wait blocks until the sync is done or canceled.
+func (s *stateSync) wait() error {
+	<-s.done
+	return s.err
+}
+
+// wait blocks until the sync is done or canceled.
+func (s *stateSync) checkDone() (bool, error) {
+	select {
+	case <-s.done:
+		return true, s.err
+	default:
+		return false, nil
+	}
+}
+
+func (s *stateSync) Cancel() error {
+	s.cancelOnce.Do(func() { close(s.cancel) })
+	return s.wait()
+}
+
+func (s *stateSync) run() {
+	s.err = s.loop()
+	close(s.done)
+}
+
+func (s *stateSync) loop() error {
+	newPeer := make(chan *peer, 200)
+	peerSub := s.d.peers.SubscribeNewPeers(newPeer)
+	defer peerSub.Unsubscribe()
+
+	for s.sched.Pending() > 0 {
+		if err := s.assignTasks(); err != nil {
+			return err
+		}
+
+		select {
+		case <-newPeer:
+			// assign new tasks
+		case <-s.cancel:
+			return errCancelStateFetch
+		case req := <-s.deliver:
+			// response or timeout
```

Comment:
This path doesn't get called on timeout, so any node which timed out will forever get stuck.